src/riscv/lib/src/machine_state/page_cache/router.rs

```diff
@@ -0,0 +1,463 @@
+    /// overlapping ranges and returns them as an iterator which the page cache can use to clean
+    /// up the invalidated pages.
+    pub fn drain_overlapping(
+        &mut self,
+        range: RangeInclusive<usize>,
+    ) -> impl Iterator<Item = RangeInclusive<usize>> {
+        let ranges = self
+            .internal
+            .overlapping(&range)
+            .map(|(r, _)| r.clone())
+            .collect::<Vec<_>>();
+        for r in &ranges {
+            self.internal.remove(r.clone());
+        }
+        ranges.into_iter()
+    }
+
+    /// Utility method that inserts an entry, either with the given `target` (it is an extension of
+    /// an existing entry, which will be overwritten) or with a new `default` target if it has
+    /// `target == None`.
+    fn insert_entry(&mut self, entry: RouterEntry<T>)
+    where
+        T: Default,
+    {
+        self.internal
+            .insert(entry.range, entry.target.unwrap_or_default());
+    }
+}
+
+/// A utility type representing an entry in a `Router` that handles the merging logic.
+struct RouterEntry<T> {
+    range: RangeInclusive<usize>,
+
+    /// The reason this is an `Option` is to represent which existing entries can be merged safely
+    /// and which new entries are actually extensions of existing entries.
+    ///
+    /// If `target` is `Some` this is an existing entry that is safe to extend by merging with a
+    /// new entry, or a new entry which is an extension of an existing entry.
+    ///
+    /// If `target` is `None` this is an existing entry that is not safe to extend, or a new entry
+    /// that will be inserted with a new `default` target.
+    target: Option<RouterTarget<T>>,
+}
+
+impl<T: Clone> RouterEntry<T> {
+    fn from_range(range: RangeInclusive<usize>) -> Self {
+        Self {
+            range,
+            target: None,
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.range.end() - self.range.start() + 1
+    }
+
+    /// Two entries are 'touching' if there is no overlap but also no gap between them.
+    ///
+    /// Assumptions:
+    ///
+    /// - `self` is before `other`.
+    fn is_touching(&self, other: &Self) -> bool {
+        self.range.end().saturating_add(1) == *other.range.start()
+    }
+
+    /// Create a `RouterEntry` for the gap between `self` and `other`. If either is safe to extend
+    /// we clone its target to the new entry; if both are safe to extend we choose the longer one,
+    /// defaulting to the left hand side (`self`).
+    ///
+    /// Assumptions:
+    ///
+    /// - `self` is before `other`,
+    /// - they are neither touching nor overlapping.
+    fn gap(&self, other: &Self) -> Self {
+        let target = match (self.target.is_some(), other.target.is_some()) {
+            (false, false) => None,
+            (true, false) => self.target.clone(),
+            (false, true) => other.target.clone(),
+            (true, true) => {
```

Comment:
```suggestion
            _ if self.len() >= other.len() => self.target.clone(),
            _ => self.target.clone()
```
as an alternative. May look a bit cleaner to avoid the nested if/else